### PR TITLE
API-8003-parser

### PR DESCRIPTION
**Pull request by jhulbert-va**
# [API-8003](https://vajira.max.gov/browse/API-8003)

### CODE REVIEW

File: charon-models/src/main/java/gov/va/api/lighthouse/charon/models/EntryParser.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package gov.va.api.lighthouse.charon.models;
+
+public interface EntryParser<E> {
```
Comment (eclendenning-va) [inline charon-models/src/main/java/gov/va/api/lighthouse/charon/models/EntryParser.java:3]: Did we want EntryParser to also Define parseNewLineDelimited?

**Review by eclendenning-va — APPROVED** [review]

File: charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+@NoArgsConstructor(staticName = "create")
+public class GetInsEntryParser implements EntryParser<GetInsEntry> {
+
+  @Override
+  public GetInsEntry parseLine(String entry) {
+    if (entry == null) {
+      throw new IllegalArgumentException("Cannot parse null entry.");
+    }
+    String[] caretSeparated = entry.split("\\^", -1);
```
Comment (bschofield-va) [inline charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java:16]: This is _acceptable_ but not _optimal_. Split on regex will (a) recompile the expression repeatedly, (b) be slow. A tokernizer approach is more code, but more performant.

**Review by bschofield-va — CHANGES_REQUESTED** [review]
hol up ... those exceptions don't seem right. lemme investigate

File: charon-models/src/test/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParserTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        arguments("1^2^3^^5", GetInsEntry.of("1", "2", "3", null, "5")),
+        arguments("1^2^3^^", GetInsEntry.of("1", "2", "3", null, null)));
+  }
+
+  static Stream<Arguments> parseLineFailureThrows() {
+    return Stream.of(
+        arguments(IllegalArgumentException.class, null),
+        arguments(IllegalArgumentException.class, "1^2^3^4"),
+        arguments(IllegalArgumentException.class, "1^2^3^4^5^6"),
+        arguments(NullPointerException.class, "^2^3^4^5"),
```
Comment (bschofield-va) [inline charon-models/src/test/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParserTest.java:27]: yeah, this doesn't seem right, these should all be through Illegal Argument.

File: charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java (hunk 2 above)
Comment (bschofield-va) [reply charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java:16]: I stand corrected. I could not implement a parser that is faster. Turns out `string.split(...)` has some interesting short circuit logic for cases just like this that make it ridiculously fast.

File: charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    if (caretSeparated.length != 5) {
+      throw new IllegalArgumentException(
+          "Invalid Number of fields. Expected 5, got " + caretSeparated.length);
+    }
+    return GetInsEntry.builder()
```
Comment (bschofield-va) [inline charon-models/src/main/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParser.java:21]:
```
    try {
      return GetInsEntry.builder()
          .fileNumber(csv.get(0))
          .ien(csv.get(1))
          .fieldNumber(csv.get(2))
          .externalValueRepresentation(csv.get(3))
          .internalValueRepresentation(csv.get(4))
          .build();
    } catch (NullPointerException e) {
      throw new IllegalArgumentException("Illegal field values: " + e.getMessage(), e);
    }
    ```

File: charon-models/src/test/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParserTest.java (hunk 3 above)
Comment (jhulbert-va) [reply charon-models/src/test/java/gov/va/api/lighthouse/charon/models/iblhsamcmsgetins/GetInsEntryParserTest.java:27]: The NPE's end up getting thrown by Lomboks `@NonNull`

**Review by eclendenning-va — APPROVED** [review]

**Review by bschofield-va — APPROVED** [review]

**Pull request merged**